### PR TITLE
[rai] improved subtitles extraction

### PR DESCRIPTION
**Pull request by nixxo**
### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
---

### Description of your *pull request* and other information

improved subtitles extraction, see #27698

### CODE REVIEW

File: youtube_dl/extractor/rai.py [hunk 1]
```diff
@@ -103,22 +103,33 @@ def _extract_relinker_info(self, relinker_url, video_id):
...
-            if subtitle_url.endswith(STL_EXT):
-                srt_url = subtitle_url[:-len(STL_EXT)] + SRT_EXT
-                subtitles['it'].append({
+            if url.endswith(STL_EXT):
+                srt_url = url[:-len(STL_EXT)] + SRT_EXT
+                sub[lang].append({
                     'ext': 'srt',
                     'url': srt_url,
                 })
+            return sub
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:122]: inline.

File: youtube_dl/extractor/rai.py [hunk 2]
```diff
@@ -103,22 +103,33 @@ def _extract_relinker_info(self, relinker_url, video_id):
...
+
+        subtitles = {}
+        subtitlesArray = video_data.get('subtitlesArray')
+        subtitlesArray.append({'url': video_data.get('subtitles')})
+        for subtitle in subtitlesArray or []:
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:127]:
```suggestion
        subtitlesArray = video_data.get('subtitlesArray') or []
        subtitlesArray.append({'url': video_data.get('subtitles')})
        for subtitle in subtitlesArray:
```

**Review by remitamine — CHANGES_REQUESTED** [review]
Add an only matching test with a URL that has `subtitlesArray`.

File: youtube_dl/extractor/rai.py [hunk 3]
```diff
@@ -103,22 +103,33 @@ def _extract_relinker_info(self, relinker_url, video_id):
...
+            if url.endswith(STL_EXT):
+                srt_url = url[:-len(STL_EXT)] + SRT_EXT
+                sub[lang].append({
                     'ext': 'srt',
                     'url': srt_url,
                 })
+            return sub
+
+        subtitles = {}
+        subtitlesArray = video_data.get('subtitlesArray')
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:125]: use the common naming convention in Python.

Comment (nixxo) [reply youtube_dl/extractor/rai.py]: fixed

File: youtube_dl/extractor/rai.py (hunk 2 above)
Comment (nixxo) [reply youtube_dl/extractor/rai.py]: fixed

File: youtube_dl/extractor/rai.py (hunk 1 above)
Comment (nixxo) [reply youtube_dl/extractor/rai.py:122]: do you mean I should remove the function `create_sub` and put the code directly in the for loop?

Comment (remitamine) [reply youtube_dl/extractor/rai.py:122]: yes, the function is used only once.

Comment (nixxo) [reply youtube_dl/extractor/rai.py]: ok, done

File: youtube_dl/extractor/rai.py [hunk 4]
```diff
@@ -103,22 +103,31 @@ def _extract_relinker_info(self, relinker_url, video_id):
...
+        for subtitle in subtitles_array:
+            sub_url = subtitle.get('url')
+            if sub_url and isinstance(sub_url, compat_str):
+                sub_lang = subtitle.get('language') or 'it'
+                sub_url = urljoin(url, sub_url)
+                sub = {}
+                sub[sub_lang] = [{
+                    'ext': determine_ext(sub_url),
+                    'url': sub_url,
+                }]
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:122]:
```suggestion
                subtitles.setdefault(sub_lang, []).append({
                    'ext': determine_ext(sub_url),
                    'url': sub_url,
                })
```

File: youtube_dl/extractor/rai.py [hunk 5]
```diff
@@ -103,22 +103,31 @@ def _extract_relinker_info(self, relinker_url, video_id):
...
-            }]
-            if subtitle_url.endswith(STL_EXT):
-                srt_url = subtitle_url[:-len(STL_EXT)] + SRT_EXT
-                subtitles['it'].append({
-                    'ext': 'srt',
-                    'url': srt_url,
-                })
+        subtitles_array = video_data.get('subtitlesArray') or []
+        subtitles_array.append({'url': video_data.get('subtitles')})
+        subtitles_array.append({'url': video_data.get('subtitlesUrl')})
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:112]:
```suggestion
        for k in ('subtitles', 'subtitlesUrl'):
            subtitles_array.append({'url': video_data.get(k)})
```

File: youtube_dl/extractor/rai.py [hunk 6]
```diff
@@ -103,22 +103,31 @@ def _extract_relinker_info(self, relinker_url, video_id):
...
+            sub_url = subtitle.get('url')
+            if sub_url and isinstance(sub_url, compat_str):
+                sub_lang = subtitle.get('language') or 'it'
+                sub_url = urljoin(url, sub_url)
+                sub = {}
+                sub[sub_lang] = [{
+                    'ext': determine_ext(sub_url),
+                    'url': sub_url,
+                }]
+                if sub_url.endswith(STL_EXT):
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:123]: use the extension extracted from `determine_ext`.

**Review by remitamine — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/rai.py [hunk 7]
```diff
@@ -103,22 +103,28 @@ def _extract_relinker_info(self, relinker_url, video_id):
...
+        for subtitle in subtitles_array:
+            sub_url = subtitle.get('url')
+            if sub_url and isinstance(sub_url, compat_str):
+                sub_lang = subtitle.get('language') or 'it'
+                sub_url = urljoin(url, sub_url)
+                subtitles.setdefault(sub_lang, []).append({
+                    'ext': determine_ext(sub_url),
+                    'url': sub_url,
                 })
+                if determine_ext(sub_url) == STL_EXT:
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:122]: DRY.

File: youtube_dl/extractor/rai.py [hunk 8]
```diff
@@ -103,22 +103,28 @@ def _extract_relinker_info(self, relinker_url, video_id):
...
+                    subtitles.setdefault(sub_lang, []).append({
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:123]: at this point `sub_lang` is guaranteed to be in the subtitles dict.

**Review by remitamine — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/rai.py [hunk 9]
```diff
@@ -103,22 +103,27 @@ def _extract_relinker_info(self, relinker_url, video_id):
...
+        for subtitle in subtitles_array:
+            sub_url = subtitle.get('url')
+            if sub_url and isinstance(sub_url, compat_str):
+                sub_lang = subtitle.get('language') or 'it'
+                sub_url = urljoin(url, sub_url)
+                subtitles.setdefault(sub_lang, []).append({
+                    'ext': determine_ext(sub_url),
+                    'url': sub_url,
                 })
+                if STL_EXT == subtitles[sub_lang][0]['ext']:
```
Comment (remitamine) [inline youtube_dl/extractor/rai.py:122]: instead of this it's better to just put result of `determine_ext` in a variable and use it.

**Pull request merged**